### ISSUE TITLE
Exceeding MaxMessageBytes causes connection to 'hang'

### ISSUE THREAD

**Issue by oogali** [emersion/go-smtp#98]
There's some odd behavior happening with MaxMessageBytes being exceeded.

```
s := smtp.NewServer(be)
s.Addr = ":2345"
s.Domain = hostname
s.AuthDisabled = true
s.MaxMessageBytes = 2
log.Fatal(server.ListenAndServe())
```

When I exceed the maximum message size, four very peculiar things happen:

1. when a new line is sent, the server immediately returns a 250 OK once the message limit has been reached
2. the connection continues to expect the terminating `<CR><LF>.<CR><LF>` sequence
3. when the sequence is sent, there is no indication the expectation has been satisfied
4. a truncated message is queued for delivery (by virtue of no error being encountered)

```
Trying ::1...
Connected to localhost.
Escape character is '^]'.
220 localhost ESMTP Service Ready
EHLO localhost
250-Hello localhost
250-PIPELINING
250-8BITMIME
250-ENHANCEDSTATUSCODES
250 SIZE 2
MAIL FROM: oogali@localhost
250 2.0.0 Roger, accepting mail from <oogali@localhost>
RCPT TO: test@localhost
250 2.0.0 I'll make sure <test@localhost> gets this
DATA
354 2.0.0 Go ahead. End your data with <CR><LF>.<CR><LF>
Hello.
250 2.0.0 OK: queued
QUIT
.
QUIT
221 2.0.0 Goodnight and good luck
Connection closed by foreign host.
```

```
$ gzcat ~/inbound-mail-queue/inboundfun.136391468.gz
From oogali@localhost  Thu May 14 01:29:25 2020
Return-Path: <oogali@localhost>
X-Original-To: test@localhost
Delivered-To: test@localhost
Received: from localhost (localhost [IPv6:::1])
	by localhost (inboundfun) with ESMTP id 8B1151CCC4
	for <test@localhost>; Thu, 14 May 2020 01:29:25 +0000 (UTC)
Message-Id: <20200514012925.8B1151CCC4@localhost>
Date: Thu, 14 May 2020 01:29:25 +0000 (UTC)
From: <oogali@localhost>

He
```

**Comment by foxcpp** [comment]
Interesting, there is a test in server_test.go that checks that it behaves correctly. Backend should receive ErrDataTooLarge on Read and handle it appropriately.

**Comment by foxcpp** [comment]
I changed MaxMessageBytes to 2 in that test and it still seems to pass. Are you able to create a self-contained example that reproduces your problem?

**Comment by oogali** [comment]
I refactored my code and I can no longer reproduce item 4, so that was on my end.

However, item 1 still occurs and I get the expected 552 response.

I can still reproduce items 2 and 3. I think it's because `textproto.DotReader` is still active in that particular context.

I don't think a 552 should be immediately generated when the limit is reached, but rather after the terminating `<CR><LF>.<CR><LF>` sequence is received.

Here's a unit test that hangs after the 552 error. It will not hang if you uncomment the `quit` scenario.

```
package inboundfun

import (
	"bufio"
	"io"
	"net"
	"testing"

	"github.com/emersion/go-smtp"
	"github.com/stretchr/testify/assert"
)

func TestServer_MessageLimit(t *testing.T) {
	l, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}

	backend, err := NewBackend("localhost")
	assert.NoError(t, err)
	assert.NotNil(t, backend)

	server := smtp.NewServer(backend)
	server.Domain = "localhost"
	server.AllowInsecureAuth = false
	server.AuthDisabled = true
	server.MaxMessageBytes = 2

	go server.Serve(l)

	client, err := net.Dial("tcp", l.Addr().String())
	if err != nil {
		t.Fatal(err)
	}

	scanner := bufio.NewScanner(client)

	t.Run("server banner", func(t *testing.T) {
		scanner.Scan()
		assert.Equal(t,"220 localhost ESMTP Service Ready", scanner.Text())
	})

	t.Run("ehlo", func(t *testing.T) {
		_, err = io.WriteString(client, "EHLO localhost\r\n")
		assert.NoError(t, err)

		scanner.Scan()
		assert.Equal(t, "250-Hello localhost", scanner.Text())

		scanner.Scan()
		assert.Equal(t, "250-PIPELINING", scanner.Text())

		scanner.Scan()
		assert.Equal(t, "250-8BITMIME", scanner.Text())

		scanner.Scan()
		assert.Equal(t, "250-ENHANCEDSTATUSCODES", scanner.Text())

		if !server.AuthDisabled && server.AllowInsecureAuth {
			scanner.Scan()
			assert.Equal(t, "250-AUTH PLAIN", scanner.Text())
		}

		scanner.Scan()
		assert.Equal(t, "250 SIZE 2", scanner.Text())
	})

	t.Run("mail from", func(t *testing.T) {
		_, err = io.WriteString(client, "MAIL FROM: test@localhost\r\n")
		assert.NoError(t, err)

		scanner.Scan()
		assert.Equal(t, "250 2.0.0 Roger, accepting mail from <test@localhost>", scanner.Text())
	})


	t.Run("rcpt to", func(t *testing.T) {
		_, err = io.WriteString(client, "RCPT TO: test@lab\r\n")
		assert.NoError(t, err)

		scanner.Scan()
		assert.Equal(t, "250 2.0.0 I'll make sure <test@lab> gets this", scanner.Text())
	})

	t.Run("data", func(t *testing.T) {
		_, err = io.WriteString(client, "DATA\r\n")
		assert.NoError(t, err)

		scanner.Scan()
		assert.Equal(t, "354 2.0.0 Go ahead. End your data with <CR><LF>.<CR><LF>", scanner.Text())

		_, err = io.WriteString(client, "Hello.\r\n")
		assert.NoError(t, err)

		scanner.Scan()
		assert.Equal(t, "552 5.3.4 Maximum message size exceeded", scanner.Text())

		/* scanner.Scan()
		assert.Equal(t, "250 2.0.0 OK: queued", scanner.Text())
		assert.Equal(t, 1, backend.queue.Length()) */
	})

	t.Run("quit", func(t *testing.T) {
		_, err = io.WriteString(client, "QUIT\r\n")
		assert.NoError(t, err)

		scanner.Scan()
		assert.Equal(t, "221 2.0.0 Goodnight and good luck", scanner.Text())
	})
}
```

**Comment by oogali** [comment]
And for comparison, here's an SMTP exchange with Postfix 3.2.2 (with `message_size_limit` clamped to 2):

```
Trying ::1...
Connected to localhost.
Escape character is '^]'.
220 supersonic.local ESMTP Postfix
EHLO localhost
250-localhost
250-PIPELINING
250-SIZE 2
250-VRFY
250-ETRN
250-ENHANCEDSTATUSCODES
250-8BITMIME
250-DSN
250 SMTPUTF8
MAIL FROM: test@localhost
250 2.1.0 Ok
RCPT TO: test@localhost
250 2.1.5 Ok
DATA
354 End data with <CR><LF>.<CR><LF>
Hello.
.
552 5.3.4 Error: message file too big
QUIT
221 2.0.0 Bye
Connection closed by foreign host.
```

**Comment by foxcpp** [comment]
>but rather after the terminating <CR><LF>.<CR><LF> sequence is received.

I would just kill connection at this point. If client continues to send multi-GiB message, this is not optimal. But I believe this decision should not be made in go-smtp (however it should allow this behavior somehow)

Instead, if Data() returns any error, go-smtp code should correctly consume remaining message contents and then send the response.

**Comment by foxcpp** [comment]
```go
	r := newDataReader(c)
	code, enhancedCode, msg := toSMTPStatus(c.Session().Data(r))
	io.Copy(ioutil.Discard, r) // Make sure all the data has been consumed
	c.WriteResponse(code, enhancedCode, msg)
```
In handleData function there is this call io.Copy call that should ensure all data is consumed before sending response.
However that io.Copy fails since dataReader implements message size limit. 

Quick solution is to adjust return type of newDataReader and set r.limited = false before calling io.Copy. This will make go-smtp server behavior match Postfix.

**Comment by foxcpp** [comment]
Your test still seems to hang with this change though...

**Comment by foxcpp** [comment]
@oogali, your test is not sending terminating `.` properly.
```diff
		scanner.Scan()
		assert.Equal(t, "354 2.0.0 Go ahead. End your data with <CR><LF>.<CR><LF>", scanner.Text())

		_, err = io.WriteString(client, "Hello.\r\n")
+		_, err = io.WriteString(client, ".\r\n")
		assert.NoError(t, err)

		scanner.Scan()
		assert.Equal(t, "552 5.3.4 Maximum message size exceeded", scanner.Text())
```
It passes with my go-smtp fix and this change to the test.

**Comment by oogali** [comment]
@foxcpp Yup, I left the terminating sequence out so I could demonstrate that the 552 was being returned before the client finished writing its data.

With regards to a fix, I briefly experimented with something on my end to respect the message size limit without exposing myself to a potential DoS (basically, do not buffer anything beyond the limit in memory).

But I didn't like that I'm now buffering the data payload in order to check the limit.

```
func (s *Session) Data(r io.Reader) error {
	msgLimit := int64(10485760)

        // create a limit reader, buffer up to $msgLimit bytes in memory... ugh
	reader := io.LimitReader(r, msgLimit)
	buf := &bytes.Buffer{}

	n, err := io.Copy(buf, reader)
	if err != nil {
		if err == io.EOF && n == msgLimit {
			return errMessageSizeExceeded
		} else {
			return err
		}
	}
	rawData := buf.Bytes()

        // attempt to read 1 more byte from the buffer, return an error if we don't get EOF
	reader = io.LimitReader(r, 1)
	n, err = io.Copy(&bytes.Buffer{}, reader)
	if err != nil && err != io.EOF {
		return err
	}

	if n > 0 {
		return errMessageSizeExceeded
	}

        ...
```

**Comment by foxcpp** [comment]
@oogali you do not need any of these with fix from PR #99.
